src/big/big_decimal.cr

```diff
@@ -237,6 +237,114 @@ struct BigDecimal < Number
+  # BigDecimal.rounding_mode = :to_zero
+  #
+  # BigDecimal.save_rounding_mode do
+  #   BigDecimal.rounding_mode = :ties_away
+  #   BigDecimal.rounding_mode # => Number::RoundingMode::TIES_AWAY
+  # end
+  #
+  # BigDecimal.rounding_mode # => Number::RoundingMode::TO_ZERO
+  # ```
+  def self.save_rounding_mode
+    prev_rounding_mode = @@rounding_mode
+    begin
+      yield
+    ensure
+      @@rounding_mode = prev_rounding_mode
+    end
+  end
+
+  # Executes the provided block, preserving the rounding mode; *mode* argument
+  # is set as `BigDecimal.rounding_mode` inside the block.
+  #
+  # ```
+  # BigDecimal.rounding_mode = :to_zero
+  #
+  # # *mode* argument is set as `BigDecimal.rounding_mode` inside the block
+  # BigDecimal.with_rounding_mode(:ties_away) do
+  #   BigDecimal.rounding_mode # => Number::RoundingMode::TIES_AWAY
+  # end
+  #
+  # BigDecimal.rounding_mode # => Number::RoundingMode::TO_ZERO
+  # ```
+  #
+  # NOTE: Uses `save_rounding_mode` internally.
+  def self.with_rounding_mode(mode : RoundingMode?)
+    save_rounding_mode do
+      @@rounding_mode = mode
+      yield
+    end
+  end
+
+  # Rounds to the nearest integer (by default), returning the result as a `BigDecimal`.
+  #
+  # ```
+  # BigDecimal.new("3.14159").round(2)    # => 3.14
+  # BigDecimal.new("8.7").round           # => 9
+  # BigDecimal.new("-9.9").round          # => -10
+  # BigDecimal.new("13345.234").round(-2) # => 13300
+  # ```
+  #
+  # If *digits* is specified and positive, the fractional part of the result
+  # has no more than that many digits.
+  #
+  # If *digits* is specified and negative, at least that many digits to the left of
+  # the decimal point will be `0` in the result.
+  #
+  # The value of the optional *mode* argument can be used to determine how
+  # rounding is performed.
+  def round(digits : Int = 0, *, mode : RoundingMode = BigDecimal.rounding_mode) : BigDecimal
+    return self if @scale <= digits
+
+    n_digits = @value.abs.digits.reverse
+    negative = @value < 0
+    scale = @scale.to_i - digits
+
+    # 123.456.to_big_d.round(-10) # => 0.1e11
+    if digits < 0 && n_digits.size <= scale
+      value = self.class.new(n_digits.first, 1)
+      value *= -1 if negative
+      value = value.round(0, mode: mode)
+      value *= power_ten_to(digits.abs)
+      return value
+    end
+
+    value = n_digits[0...-scale].join.presence.try(&.to_big_i) || 0
+    value = self.class.new(value, digits < 0 ? 0 : digits)
+
+    # 10001.123.to_big_d.round(-3) # => 10000
+    msd = n_digits[-scale]
+    if msd == 0
+      value *= -1 if negative
+      value *= power_ten_to(digits.abs) if digits < 0
+      return value
+    end
+
+    round_up =
+      case mode
+      in .to_zero?     then false
+      in .ties_away?   then msd >= 5
+      in .ties_even?   then msd == 5 ? n_digits[-2]?.try(&.odd?) : msd > 5
+      in .to_positive? then !negative
+      in .to_negative? then negative
+      end
+
+    value += 1 if round_up
```

Comment:
This addition is incorrect, since `value` is already the `BigDecimal` itself, not the mantissa:

```crystal
BigDecimal.new("123.456789").round(2, mode: :ties_away) # => 124.45
```